File: 2019/Intcode.rb
class Intcode

  NUM_PARAMS = {
    1 => 3,
    2 => 3,
    3 => 1,
    4 => 1,
    5 => 2,
    6 => 2,
    7 => 3,
    8 => 3,
    9 => 1,
    99 => 0
  }

  attr_accessor :id, :input, :output, :is_running
  def initialize(program, input, output, id)
    @id = id
    @input = input
    @output = output
    @instruction_pointer = 0
    @relative_base = 0
    @program = (0...program.length).zip(program).to_h
    @program.default = 0
    @is_running = false
  end

  def get_params()
    instruction = @program[@instruction_pointer]
    params = []
    (1..NUM_PARAMS[instruction%100]).map do |i|
      mode = (instruction / (10**(i+1))) % 10
      next @program[@instruction_pointer+i] if mode==0 # position mode
      next @instruction_pointer+i if mode==1 # immediate mode
      next @relative_base+@program[@instruction_pointer+i] if mode==2 # relative mode
    end
  end

  def run
    @is_running = true
    all_outputs = []
    loop do
      opcode = @program[@instruction_pointer] % 100
      params = get_params()

      if opcode == 1 # add
        @program[params[2]] = @program[params[0]] + @program[params[1]]
      elsif opcode == 2 # multiply
        @program[params[2]] = @program[params[0]] * @program[params[1]]
      elsif opcode == 3 # input
        sleep 0.01 while @input.empty?
        @program[params[0]] = @input.shift
      elsif opcode == 4 # output
        all_outputs << @program[params[0]]
        @output << @program[params[0]]
      elsif opcode == 5 # jump-if-true
        next (@instruction_pointer = @program[params[1]]) if @program[params[0]] != 0
      elsif opcode == 6 # jump-if-false
        next (@instruction_pointer = @program[params[1]]) if @program[params[0]] == 0
      elsif opcode == 7 # less than
        @program[params[2]] = @program[params[0]] < @program[params[1]] ? 1 : 0
      elsif opcode == 8 # equals
        @program[params[2]] = @program[params[0]] == @program[params[1]] ? 1 : 0
      elsif opcode == 9
        @relative_base += @program[params[0]]
      elsif opcode == 99
        @is_running = false
        return
      end
      @instruction_pointer += params.length + 1
    end
  end
end

File: 2019/elixir/lib/intcode.ex
defmodule Aoc2019.Intcode do
  use GenServer
  import Enum
  import Aoc2019.Intcode.Opcode
  alias Aoc2019.Intcode.Opcode

  defmodule State do
    defstruct(
      memory: %{},
      ptr: 0,
      input: [],
      output: [],
      status: :on,
      relative_base: 0
    )
  end

  ## Client

  def new(program, input \\ []) do
    memory = program |> with_index() |> map(fn {v, i} -> {i, v} end) |> Map.new()
    GenServer.start_link(__MODULE__, %State{memory: memory, input: input})
  end

  def run(vm), do: GenServer.cast(vm, :run)
  def run_sync(vm), do: GenServer.call(vm, :run)
  def step(vm), do: GenServer.call(vm, :step)
  def input(vm, n), do: GenServer.cast(vm, {:input, n})
  def flush_output(vm), do: GenServer.call(vm, :flush_output)
  def get_state(vm), do: GenServer.call(vm, :get_state)

  ## Server

  @impl true
  def init(state), do: {:ok, state}

  @impl true
  def handle_call(:get_state, _from, state) do
    {:reply, state, state}
  end

  @impl true
  def handle_call(:flush_output, _from, state) do
    {:reply, state.output, %{state | output: []}}
  end

  @impl true
  def handle_call(:run, _from, state) do
    state = _run(state)

    if state.status == :halted,
      do: {:stop, :normal, state, state},
      else: {:reply, state, state}
  end

  @impl true
  def handle_call(:step, _from, state) do
    state = _step(state)

    if state.status == :halted,
      do: {:stop, :normal, state, state},
      else: {:reply, state, state}
  end

  @impl true
  def handle_cast(:run, state) do
    state = _run(state)

    if state.status == :halted,
      do: {:stop, :normal, state},
      else: {:noreply, state}
  end

  @impl true
  def handle_cast({:input, n}, state) do
    lst = if is_list(n), do: n, else: [n]
    {:noreply, Map.update!(state, :input, &(&1 ++ lst))}
  end

  defp _run(%{status: :halted} = state), do: state
  defp _run(%{status: :waiting_for_input, input: []} = state), do: state
  defp _run({:unrecognized_opcode, opcode}), do: {:unrecognized_opcode, opcode}
  defp _run(state), do: _step(state) |> _run()

  defp _step(%{memory: memory, ptr: ptr} = state) do
    instruction = Map.get(memory, ptr) |> parse_instruction()

    case instruction.opcode do
      99 -> halt(state)
      1 -> add(instruction.param_modes, state)
      2 -> mult(instruction.param_modes, state)
      3 -> Opcode.input(instruction.param_modes, state)
      4 -> Opcode.output(instruction.param_modes, state)
      5 -> jump_if_true(instruction.param_modes, state)
      6 -> jump_if_false(instruction.param_modes, state)
      7 -> less_than(instruction.param_modes, state)
      8 -> equals(instruction.param_modes, state)
      9 -> adjust_relative_base(instruction.param_modes, state)
      opcode -> {:unrecognized_opcode, opcode}
    end
  end

  defp opcode(val), do: rem(val, 100)

  defp param_modes(val) do
    lookup = %{0 => :position, 1 => :immediate, 2 => :relative}
    (val / 100) |> floor() |> Integer.digits() |> reverse() |> map(fn n -> lookup[n] end)
  end

  defp parse_instruction(instruction),
    do: %{opcode: opcode(instruction), param_modes: param_modes(instruction)}
end
